sys/kern/event.c

```diff
@@ -1,15 +1,452 @@
+ *
+ * `kn_objlock` (the lock protecting the object) must be taken BEFORE `kq_lock`.
+ */
+typedef struct kqueue {
+  int kq_count;          /* number of pending events */
+  knote_tailq_t kq_head; /* list of pending events */
+  mtx_t kq_lock;         /* mutex for queue access */
+  condvar_t kq_cv;
+  knlist_t kq_knhash[KN_HASHSIZE]; /* hash table for knotes */
+} kqueue_t;
+
+/* Returns a hash bucket for a given object */
+static inline knlist_t *kq_gethash(kqueue_t *kq, void *obj) {
+  return &kq->kq_knhash[((unsigned long)obj) % KN_HASHSIZE];
+}
+
+static kqueue_t *kqueue_create(void) {
+  kqueue_t *kq = kmalloc(M_DEV, sizeof(kqueue_t), M_ZERO);
+  mtx_init(&kq->kq_lock, 0);
+  cv_init(&kq->kq_cv, 0);
+
+  TAILQ_INIT(&kq->kq_head);
+  for (int i = 0; i < KN_HASHSIZE; i++)
+    SLIST_INIT(&kq->kq_knhash[i]);
+
+  return kq;
+}
+
+/* Disconnects all knotes from the given kqueue. */
+static void kqueue_drain(kqueue_t *kq) {
+  knote_t *kn;
+
+  for (int i = 0; i < KN_HASHSIZE; i++) {
+    while ((kn = SLIST_FIRST(&kq->kq_knhash[i])) != NULL) {
+      knote_drop(kn);
+    }
+  }
+}
+
+static void kqueue_destroy(kqueue_t *kq) {
+  kqueue_drain(kq);
+  mtx_destroy(&kq->kq_lock);
+  cv_destroy(&kq->kq_cv);
+  kfree(M_DEV, kq);
+}
+
+static int filt_fileattach(knote_t *kn) {
+  file_t *fp = kn->kn_obj;
+  return fp->f_ops->fo_kqfilter(fp, kn);
+}
+
+/* Here we only need to define filt_attach function, as the rest will be
+ * specified in filterops of a concrete file type.
+ */
+static filterops_t file_filtops = {
+  .filt_attach = filt_fileattach,
+};
+
+static filterops_t *sys_kfilters[EVFILT_SYSCOUNT] = {
+  [EVFILT_READ] = &file_filtops,
+  [EVFILT_WRITE] = &file_filtops,
+};
+
+static filterops_t *filt_getops(uint32_t filter) {
+  if (filter > EVFILT_SYSCOUNT)
+    return NULL;
+  return sys_kfilters[filter];
+}
+
+static int kqueue_read(file_t *f, uio_t *uio) {
+  return EOPNOTSUPP;
+}
+
+static int kqueue_write(file_t *f, uio_t *uio) {
+  return EOPNOTSUPP;
+}
+
+static int kqueue_close(file_t *f) {
+  kqueue_t *kq = f->f_data;
+  kqueue_destroy(kq);
+  f->f_data = NULL;
+
+  return 0;
+}
+
+static int kqueue_stat(file_t *f, stat_t *sb) {
+  return EOPNOTSUPP;
+}
+
+static int kqueue_seek(file_t *f, off_t offset, int whence, off_t *newoffp) {
+  return EOPNOTSUPP;
+}
+
+static int kqueue_ioctl(file_t *f, u_long cmd, void *data) {
+  return EOPNOTSUPP;
+}
+
+static fileops_t kqueueops = {.fo_read = kqueue_read,
+                              .fo_write = kqueue_write,
+                              .fo_close = kqueue_close,
+                              .fo_stat = kqueue_stat,
+                              .fo_seek = kqueue_seek,
+                              .fo_ioctl = kqueue_ioctl};
+
+static int kqueue_get_obj(proc_t *p, kevent_t *kev, void **obj) {
+  if (kev->filter == EVFILT_READ || kev->filter == EVFILT_WRITE)
+    return fdtab_get_file(p->p_fdtable, kev->ident, FF_READ, (file_t **)obj);
+
+  return EINVAL;
```

Comment:
According to NetBSD and OpenBSD manuals, a bad file descriptor should cause `kevent` to return `EBADF`.

```suggestion
  return EBADF;
```

Reply:
This error code is returned when a filter is invalid, not a file description. And such a case we should return `EINVAL`.

An excerpt from the manual:
`   [EINVAL]		The specified time limit or filter is invalid.`